tensorflow/core/kernels/stack.cc

```diff
@@ -115,6 +191,12 @@ class Stack : public ResourceBase {
   bool closed_ GUARDED_BY(mu_);
   std::vector<TensorAndAllocation> stack_ GUARDED_BY(mu_);
 
+  std::queue<int> unswapped_ids_ GUARDED_BY(mu_);
```

Comment:
Instead of a queue, I'd prefer if we kept this as a std::list and the main stack is a std::pair<TensorAndAllocation, std::list::iterator list_element>>

And push/pop erase from that list. That way the stack is consistent with the list at all times, and we don't have to worry about the index not being correct. In this implementation, it's possible for them to get out of sync.

Reply:
Thank you for the response. I've been thinking this through but, sorry, I'm not sure I get what you mean.

Do you mean that `stack`, `unswapped_ids`, and `swapped_ids` should be like this?

```
  std::vector<std::pair<TensorAndAllocation, std::list<int>::iterator>> stack_ GUARDED_BY(mu_);

  std::list<int> unswapped_ids_ GUARDED_BY(mu_);
  std::list<int> swapped_ids_ GUARDED_BY(mu_);
```

 If so, I still don't quite understand the intention of pairing stack elements with list iterators to `unswapped_ids`. Could you please elaborate more?

And about them being out of sync, I might be missing something here, but here's my explanation why I think they won't:

1. In this implementation, we swap out the tensor with `id` popped from the front of `unswapped_ids`. Afterward, that `id` is pushed to the back of `swapped_ids`. These can be seen in `GetTensorToSwapOut`. Therefore, `unswapped_ids` and `swapped_ids` are in sync with which tensors in the stack are swapped out.

2. In the case of swapping in, at `GetTensorToSwapIn` there's this piece of code:

```
      *index = swapped_ids_.top();
      swapped_ids_.pop();
      if (*index >= stack_.size()) {
        continue;
      }
```

which ensures that if the tensor referred by `id` popped from `swapped_ids` is no longer in the stack, it's skipped.

Also, in `Stack.Pop()`, we ensure that if a tensor to be popped from the stack is still being swapped in, it waits until it's done, as shown here:

```
    while (is_swapping_ins_[index]) {
      cond_swapping_ins_[index].wait(l);
      index = stack_.size() - 1;
    }
```

Please let me know if I misunderstood your response. Also, thanks for the suggestion about benchmarking with that test, I'm currently running it.

Reply:
My proposal is something similar to how an LRU cache works, where you keep a list around to keep track of the least recently used item, but make sure they're in sync at all times. Whenever an item is added or removed from an LRU cache, both the map and the list are kept in sync. The logic that way is easier to verify when they're always consistent.

It would look like this:
```
std::list<int> unswapped_lru_ GUARDED_BY(mu_);  // holds indices into stack_
std::vector<std::pair<TensorAndAllocation, std::list<int>::iterator>> stack_ GUARDED_BY(mu_);
```

Stack::Push will push_back() onto stack_ and push_back() the index onto unswapped_lru_.
Stack::Pop will pop from stack_ and call erase() on the iterator for that element from unswapped_lru. Basically upon Stack::Pop, unswapped_lru_ will always be consistent with the contents of stack_, which is the difference from the current code.

When we need to swap memory to CPU, we pop_front() from unswapped_lru_ and swap that tensor over.

Reply:
Thank you for the clear explanation. I've pushed the latest change, is it in line with what you proposed?

However, now I'm getting errors (core dumped). I suspect it has to do with wrong iterator handling, particularly around this code in `Stack.Pop()`:

```
    *value = stack_.back().first;
    auto iter = stack_.back().second;
    if (iter != unswapped_lru_.end()) {
      unswapped_lru_.erase(iter);
    }
    stack_.pop_back();
```

To be honest, I'm not very familiar with C++ iterators, so your help will be appreciated, thanks.